File: apps/passenger-android/src/main/kotlin/lk/mageride/passenger/history/TripDetailsScreen.kt
package lk.mageride.passenger.history

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.ArrowBack
import androidx.compose.material.icons.filled.Flag
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.res.stringResource
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import lk.mageride.passenger.R
import lk.mageride.passenger.map.MageRideMap
import lk.mageride.passenger.map.MapCamera
import lk.mageride.passenger.ui.MoneyFormat
import lk.mageride.passenger.ui.component.InlineError
import lk.mageride.passenger.ui.component.SectionLabel
import lk.mageride.passenger.ui.theme.MageRideTheme

/**
 * SCR-PA-023 — one trip, its route and its receipt.
 *
 * The wireframe: the map with the trip's track, then Date / Vehicle / Driver / Distance / Total,
 * then `⬇ Receipt` and `⚑ Report issue`.
 *
 * **The distance shown is the one the fare was computed from** — `TripDetail.distanceKm`, the
 * Kalman-filtered figure (E-04) — and never a sum over the decoded polyline. When the geometry came
 * from the 1/min operational sample rather than full telemetry the contract says the distance is a
 * **lower bound**, and the screen says so rather than presenting a floor as a measurement.
 */
@Composable
internal fun TripDetailsScreen(onBack: () -> Unit, onReportIssue: () -> Unit, model: TripDetailsViewModel) {
    val state by model.state.collectAsStateWithLifecycle()

    Column(modifier = Modifier.fillMaxSize()) {
        Row(modifier = Modifier.fillMaxWidth(), verticalAlignment = Alignment.CenterVertically) {
            IconButton(onClick = onBack) {
                Icon(
                    imageVector = Icons.AutoMirrored.Filled.ArrowBack,
                    contentDescription = stringResource(R.string.action_back),
                    tint = MaterialTheme.colorScheme.onSurfaceVariant,
                )
            }
            Text(
                text = stringResource(R.string.trip_details_title),
                style = MaterialTheme.typography.titleLarge,
                color = MaterialTheme.colorScheme.onSurface,
            )
        }

        Box(modifier = Modifier.weight(MAP_WEIGHT)) {
            MageRideMap(
                modifier = Modifier.fillMaxSize(),
                routePolyline = state.route,
                camera = state.route.firstOrNull()?.let { MapCamera(it.lat, it.lng) } ?: MapCamera.Default,
            )
        }

        Column(
            modifier = Modifier
                .weight(SHEET_WEIGHT)
                .fillMaxWidth()
                .padding(MageRideTheme.spacing.md),
            verticalArrangement = Arrangement.spacedBy(MageRideTheme.spacing.xs),
        ) {
            SectionLabel(text = stringResource(R.string.trip_details_receipt))

            state.trip?.let { trip ->
                DetailRow(stringResource(R.string.trip_details_driver), trip.driver?.name.orEmpty())
                DetailRow(
                    stringResource(R.string.trip_details_vehicle),
                    trip.driver?.registrationNumber.orEmpty(),
                )
                DetailRow(
                    stringResource(R.string.trip_details_distance),
                    trip.distanceKm?.let { stringResource(R.string.summary_km, it) } ?: MoneyFormat.EMPTY,
                )
                DetailRow(
                    stringResource(R.string.summary_total),
                    trip.fareMinor?.let(MoneyFormat::rupees) ?: MoneyFormat.EMPTY,
                )
            }

            // The contract's own caveat, surfaced. A lower bound presented as a measurement is how
            // a passenger ends up arguing with a receipt that was never claiming to be exact.
            if (state.approximate) {
                Text(
                    text = stringResource(R.string.trip_details_approximate),
                    style = MaterialTheme.typography.labelSmall,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                )
            }

            state.error?.let { InlineError(message = stringResource(it)) }

            OutlinedButton(onClick = onReportIssue, modifier = Modifier.fillMaxWidth()) {
                Icon(imageVector = Icons.Filled.Flag, contentDescription = null)
                Text(
                    text = stringResource(R.string.trip_details_report),
                    modifier = Modifier.padding(start = MageRideTheme.spacing.xxs),
                )
            }
        }
    }
}

@Composable
private fun DetailRow(label: String, value: String) {
    Row(modifier = Modifier.fillMaxWidth(), horizontalArrangement = Arrangement.SpaceBetween) {
        Text(
            text = label,
            style = MaterialTheme.typography.bodyMedium,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
        )
        Text(
            text = value,
            style = MaterialTheme.typography.bodyMedium,
            color = MaterialTheme.colorScheme.onSurface,
        )
    }
}

private const val MAP_WEIGHT = 0.45f
private const val SHEET_WEIGHT = 0.55f

File: apps/passenger-ios/PassengerApp/History/TripDetailsScreen.swift
import MageRideShared
import SwiftUI

/// SCR-PI-023 — one trip, its route and its receipt.
///
/// The cell: `‹ Trips · Trip details`, a short rounded map with the track on it, a `glist` of
/// Date / Vehicle / Driver / Total, and a `btn-row` of `⬇ Receipt` and `⚑ Report`.
///
/// **The distance shown is the one the fare was computed from** — `TripDetail.distanceKm`, the
/// Kalman-filtered figure (E-04) — and never a sum over the decoded polyline. When the geometry came
/// from the 1/min operational sample rather than full telemetry the contract calls that distance a
/// **lower bound**, and the screen says so rather than presenting a floor as a measurement. The row
/// is drawn because `passenger_android.html`'s cell for the same `KEEP` screen draws it and
/// `apps/passenger-android` shows it; the iOS cell's four-row `glist` omits it, which is a difference
/// between the two drawings rather than between the two apps (C099 handoff).
///
/// **`⬇ Receipt` is drawn and disabled, and that is a contract gap rather than an omission.** No
/// passenger-facing operation on the app surface produces a trip receipt: `public-bff.yaml`'s
/// `GET /public/track/{token}/receipt` is the *web tracking token's* and `fleet-billing.yaml`'s is a
/// fleet invoice's. So the control is drawn where the wireframe draws it, refuses the tap, and says
/// why — the same call C095 made on SCR-PI-004's avatar badge, which draws an upload the platform has
/// no route for. `apps/passenger-android` omits the control entirely; recorded in the C099 handoff.
@MainActor
struct TripDetailsScreen: View {

    @StateObject private var model: TripDetailsModel

    /// `⚑ Report` — SCR-PI-030.
    let onReportIssue: () -> Void

    init(
        tripId: String,
        history: HistoryRepository,
        sessions: PassengerSessions,
        onReportIssue: @escaping () -> Void
    ) {
        _model = StateObject(wrappedValue: TripDetailsModel(tripId: tripId, history: history, sessions: sessions))
        self.onReportIssue = onReportIssue
    }

    var body: some View {
        ScrollView {
            VStack(spacing: MageRideSpacing.sm) {
                map
                receipt

                if model.state.isApproximate {
                    // The contract's own caveat, surfaced. A lower bound presented as a measurement
                    // is how a passenger ends up arguing with a receipt that never claimed to be
                    // exact.
                    Text(key: "trip_details_approximate")
                        .mageFont(.caption)
                        .foregroundStyle(MageRideColor.onSurfaceVariant)
                        .frame(maxWidth: .infinity, alignment: .leading)
                }

                if let errorKey = model.state.errorKey {
                    FormErrorText(messageKey: errorKey)
                }

                actions
            }
            .padding(MageRideSpacing.md)
        }
        .background(MageRideColor.surface)
        .navigationTitle(Text(key: "trip_details_title"))
        .navigationBarTitleDisplayMode(.inline)
        .task { await model.load() }
    }

    // MARK: -

    /// MAP-08's trip line. A track with fewer than two points draws nothing, which is what a Mode C
    /// ride whose `geometrySource` is `none` actually has.
    private var map: some View {
        MageRideMap(routePolyline: model.state.route, camera: camera)
            .allowsHitTesting(false)
            .frame(height: MageRideControl.tripDetailMapHeight)
            .clipShape(RoundedRectangle(cornerRadius: MageRideRadius.md, style: .continuous))
    }

    /// The cell's `.glist` — four rows and an emphasised total.
    ///
    /// Every row draws the dash rather than disappearing when its value is absent: a receipt with a
    /// missing line reads as a receipt that is still loading, and each of these is missing for a
    /// documented reason (see ``TripDetailsState``).
    private var receipt: some View {
        VStack(spacing: MageRideSpacing.xs) {
            KeyValueRow(titleKey: "trip_details_date", value: model.state.startedAt ?? MageRideSymbols.unknown)
            KeyValueRow(titleKey: "trip_details_vehicle", value: model.state.vehicle ?? MageRideSymbols.unknown)
            KeyValueRow(titleKey: "trip_details_driver", value: model.state.driver ?? MageRideSymbols.unknown)
            KeyValueRow(titleKey: "trip_details_distance", value: model.state.distance ?? MageRideSymbols.unknown)
            KeyValueRow(titleKey: "summary_total", value: model.state.total ?? MoneyFormat.pending, isTotal: true)
        }
        .padding(MageRideSpacing.sm)
        .background(
            MageRideColor.background,
            in: RoundedRectangle(cornerRadius: MageRideRadius.md, style: .continuous)
        )
        .redacted(reason: model.state.isLoading ? .placeholder : [])
    }

    /// The cell's `btn-row`, and the one sentence that stops the left-hand button being a mystery.
    private var actions: some View {
        VStack(spacing: MageRideSpacing.xs) {
            HStack(spacing: MageRideSpacing.xs) {
                // Drawn, refused, and explained underneath. See the type's note.
                OutlinedAction(
                    titleKey: "trip_details_receipt",
                    symbolName: "arrow.down.circle",
                    tint: MageRideColor.outline
                ) { }
                    .disabled(true)

                OutlinedAction(titleKey: "trip_details_report", symbolName: "flag.fill", action: onReportIssue)
            }

            Text(key: "trip_details_receipt_unavailable")
                .mageFont(.caption)
                .foregroundStyle(MageRideColor.onSurfaceVariant)
                .frame(maxWidth: .infinity, alignment: .leading)
        }
    }

    /// Opens on the start of the track, or on the pickup when there is no track at all.
    private var camera: MapCamera {
        if let first = model.state.route.first {
            return MapCamera(lat: first.lat, lng: first.lng)
        }
        guard let pickup = model.state.trip?.pickup else { return .colombo }
        return MapCamera(lat: pickup.lat, lng: pickup.lng)
    }
}
